api/types/replacement.go

```diff
@@ -6,22 +6,47 @@ package types
 // Replacement defines how to perform a substitution
 // where it is from and where it is to.
 type Replacement struct {
-	Source *ReplSource `json:"source" yaml:"source"`
-	Target *ReplTarget `json:"target" yaml:"target"`
+	// Raw value of replacement
+	SourceRaw string `json:"sourceRaw" yaml:"sourceRaw"`
+
+	// A specific field to read it from.
+	// Either SourceRaw or SourceField can be specified
+	SourceField *SelectedField `json:"sourceField" yaml:"sourceField"`
+
+	// The N fields to write the value to.
+	Targets []*SelectedField `json:"targets" yaml:"targets"`
 }
 
-// ReplSource defines where a substitution is from
-// It can from two different kinds of sources
-//  - from a field of one resource
-//  - from a string
-type ReplSource struct {
-	ObjRef   *Target `json:"objref,omitempty" yaml:"objref,omitempty"`
-	FieldRef string  `json:"fieldref,omitempty" yaml:"fiedldref,omitempty"`
-	Value    string  `json:"value,omitempty" yaml:"value,omitempty"`
+// SelectedField specifies one field in one or more objects.
+// Context determines if multiple objects are allowed.
+type SelectedField struct {
+	// Allow objects that match this.
+	Allow *Selector `json:"allow" yaml:"allow"`
+
+	// From the allowed set, remove objects that match this.
+	// TODO (#3492): Remove matches listed in the deny field
+	//  Currently this field is unused
+	Deny *Selector `json:"deny" yaml:"deny"`
+
+	// Structured field path expected in each allowed object.
+	FieldPath string `json:"fieldPath" yaml:"fieldPath"`
+
+	// Structured field paths expected in each allowed object.
+	// This is only allowed for targets, and only one of `fieldPath`
+	// or `fieldPaths` is allowed
+	FieldPaths []string `json:"fieldPaths" yaml:"fieldPaths"`
```

Comment:
> This is only allowed for targets

Related to my comment above, maybe the API would be cleaner if we didn't try to make source and target share this struct. E.g. source could just have an embedded selector, a singular field path and field options. Just a thought. Maybe I'm missing something about the requirements.

As for targets, do we need both singular and plural? `fieldPaths: [spec.template.spec.containers]` isn't much worse to write than `fieldPath: spec.template.spec.containers`, and the user will need to understand whether the fields are mutually exclusive or merged. 🤔

Reply:
Something like this?
```
type Kustomization struct {
  ...
  Replacments []*Replacement
  ...
}

type Replacement struct {
  Source       *ReplSource       // The source of the value.
  Targets      []*TargetSelector // The N fields to write the value to.
}

type ReplSource struct {
  SourceRaw    string                   // Raw value of replacement OR
  SourceField  *SourceSelector  // a specific field to read it from.
}

type SourceSelector struct {
  Allow     *Selector               // Allow objects that match this.
  Deny      *Selector              // From the allowed set, remove objects that match this.
  FieldPath string                  // Structured field path expected in each allowed object.
  Options   *FieldOptions     // Options that refine interpretation of the field.
}

type TargetSelector struct {
  Allow     *Selector                // Allow objects that match this.
  Deny      *Selector               // From the allowed set, remove objects that match this.
  FieldPaths []string              // Structured field paths expected in each allowed object.
  Options   *FieldOptions     // Options that refine interpretation of the field.
}

type FieldOptions struct {
  Delimiter  string            // Used to split/join the field.
  Index      int                   // Which position in the split to consider.
}
```

@monopole wdyt? I think it would be fine. It shows the differences between source and targets more clearly

Reply:
Yeah, that looks like what I was imagining. I have a new question though: if the user specifies multiple field paths in a target, how do the field options interact with those paths? Don't they need to be field-specific?

Reply:
If the user specifies multiple field paths in a target, the idea is that the value should be replaced in all of those field paths. https://github.com/kubernetes-sigs/kustomize/blob/b920d64adfbc38a2d471d0d7f3342f8c3b2c4fc0/api/filters/replacement/replacement_test.go#L190 is an example of what that might look like

Reply:
Yeah, totally, that make sense. What I'm wondering about is the not-yet-implemented `FieldOptions`, which is described as refining the interpretation of "the field"... but what does that mean when there are multiple fields? Do we need to make it so that each FieldPath can be (optionally) paired with a FieldOptions?

Reply:
So overall what users would write would looks something like this?

```yaml
replacements:
# minimal
 - source:
     value: "foo"
   targets:
   - include:
       kind: Deployment
     fields:
     - path: spec.template.spec.containers.0.args.1
# all options
 - source:
     objectRef:
       kind: ConfigMap
       name: cm
     field:
       path: data.HOSTNAME
       delimeter: "/"
       index: 0
   targets:
   - include:
       kind: Deployment
     fields:
     - path: spec.template.spec.containers.0.args.0
     - path: spec.template.spec.containers.1.args.1 
       delimeter: ":"
       index: 0
```

Reply:
Yes, that is what it would look like if we decide on this config

Reply:
Yes, that is what it would look like if we decide on this config

Reply:
It does bug me a bit how many lines it takes to write out my "minimal" example, but I don't think we can cut it down without limiting features. I think it is clear, at least. LGTM.

Reply:
If the source alternatives could both be used at the same time, they'd certainly belong in the same struct.  But since they are mutually exclusive, it's less appealing.  You get a perpetual 'hanging field', e.g.
```
source:
  raw: foo
target:
  ...
```
instead of
```
sourceRaw: foo
target:
  ...
```
Don't feel too strong about it though.  Perhaps the former form is more readable since `target` is always going to have the subfields below it.  Up to you Natasha.

---

As to always combining `FieldOptions`  with `FieldPath` in a new struct, i'd argue it's too flexible and error prone.

This only impacts cases where there are multiple FieldPaths in the target.  I cannot think of a common case where, in one replacement to multiple field paths, that one would interpret those fields differently, e.g.  treat one field as a file path and stick the replacement value into the 3rd path position, and in some other field, put the replacement value in the 1st position (or don't treat it as a path at all).  

Seems best to encourage use of one set of FieldOptions per replication, rather than obligating the user to copy it N times most of the time. 

If a different FieldOption set is needed, the user can fall back to adding another Replacement stanza - repeating the source and target kind, which is why this case will be more wordy.

---

I'd use `Include` in both source and target, rather than ObjRef in one and Include in the other. Signal that these function the same way.

---

I prefer select/reject to include/exclude, since _select_ has meaning in kubernetes and we're using it in that sense.

Either set is better than the allow/deny i initially suggested.

Reply:
Also agree with @KnVerey that we might as well omit the singular `FieldPath`

Here's an argument to keep `SourceRaw` as a distinct field, and not immediately offer it.

We're replacing `vars`, and vars always get their replacement value from some field in some object in the data set (e.g a configmap or a namespace object or a deployment) - rather than from a raw value in the kustomization file.  For this transformer, the source value doesn't have to be a simple string - it can be the pod spec (yaml subtree) in a Deployment, which we want to copy into a CronJob.

We don't want the kustomization file to morph into a helm chart style _values file_ (since helm exists).  The base should be working yaml, and not require a replacement to look correct.  In practice, things like a project ID will have to be some dummy value e.g. 100100100, but a type correct plaecholder value is better than using something that's type incorrect like `$(PROJECT_ID)`.

So we want to encourage
```
source:
 select:
 fieldPath:
targets:
- select:
  fieldPaths:
- select:
  fieldPaths:
```

and defer this use:
```
sourceRaw:  foo
targets:
 - ...
 - ...
```

This latter use case is best suited for a _setter_.  So we can defer adding `SourceRaw` immediately, and focus instead on making sure kustomize has documented support and examples for setter use.

It's easy to add `SourceRaw` later if it makes sense, but very hard to remove it.


So all that said, how about:
```
type Replacement struct {
  // Nope: SourceRaw  string
  Source     *SourceSelector   // Specific field to get data from
  Targets    []*TargetSelector // Where to write the value.
}

type SourceSelector struct {
  Select    *Selector          // Should match 1 resource to read for the replacement value.
  FieldPath  string            // Structured field path to read in that resource.
  Options    *FieldOptions     // Options that refine interpretation of the field.
}

type TargetSelector struct {
  Select     *Selector         // Select target objects that match this.
  Reject     *Selector         // From the selected set, reject objects that match this.
  FieldPaths []string          // Structured field paths to write (if present) in each allowed object.
  Options    *FieldOptions     // Options that refine interpretation of all fields in this target.
}

type FieldOptions struct {
  Delimiter       string       // Used to split/join the field.
  Index           int          // Which position in the split to consider.
  Encoding        string       // None, Base64, URL, Hex, etc.
  CreateIfMissing bool         // If field missing, add it.
}
```

Reply:
Btw i just tossed more stuff into the FieldOptions.  we can defer all of it for now.

Reply:
> cannot think of a common case where, in one replacement to multiple field paths, that one would interpret those fields differently, e.g. treat one field as a file path and stick the replacement value into the 3rd path position, and in some other field, put the replacement value in the 1st position (or don't treat it as a path at all).

While this may not be the intent of the feature, it is so powerful that I can definitely imagine a way to use it like that. For example, perhaps you have an overlay for your prod environment, and you want to inject "production" in all your Deployments in a bunch of differently structured places, e.g. as a suffix on the name, as a label, in an environment variable, and in a domain name within another environment variable, and so on. It would make sense to want to group such things together I think. Is that something we want to enable? 

> I'd use Include in both source and target, rather than ObjRef in one and Include in the other. Signal that these function the same way.

They kinda do, but the fact that one targets a single resource and the other selects a group is significant. To me, "select" sounds like it potentially expects multiple results. Related to this, it looks like vars only supported a specific object reference. Related to your point about the raw field, do we need to support selection in source? Is it going to lead us to need to support rejection as well?

Reply:
> While this may not be the intent of the feature, it is so powerful that I can definitely imagine a way to use it like that. For example, perhaps you have an overlay for your prod environment, and you want to inject "production" in all your Deployments in a bunch of differently structured places, e.g. as a suffix on the name, as a label, in an environment variable, and in a domain name within another environment variable, and so on. It would make sense to want to group such things together I think. Is that something we want to enable?

Based on recent comments I am inclined to agree with @monopole. I think this case would be less common than someone wanting the same set of field options on each target. It would make sense for the more common use case to be less wordy. 

> They kinda do, but the fact that one targets a single resource and the other selects a group is significant. To me, "select" sounds like it potentially expects multiple results. 

I have no strong opinion on this but I think with sufficient documentation it should be okay. For now I will remove sourceRaw and change include and objref to select but can update the PR if there is further discussion. 

> Is it going to lead us to need to support rejection as well?
What would be the use case for this?

Reply:
The use case for rejection would be the same thing as in the targets field: if targeting is based on selection, you may want rejection to refine the returned set down to what you actually wanted. There's also a consistency argument if we're using the same field names in both places. Still curious why we want selection at all though, given that it needs to resolve to a single resource, and this was identifier-based in the feature we're replacing.

Reply:
We could require apiVersion, kind, and name for a source to make it identifier-based but personally I don't see why it would be necessary so long as we require source to resolve to a single resource.

Reply:
w/r `SourceSelector`, i agree that avoiding the possibility of a multiple hits in the source trumps the 'elegance' of reusing `Selector`.  So how about:
```
type SourceSelector struct {
  resid.Gvk
  Name string
  Namespace  string 
  FieldPath  string            // Structured field path to read in that resource.
  Options    *FieldOptions     // Options that refine interpretation of the field.
}
```
Compare this to 
- `Target` in: https://github.com/kubernetes-sigs/kustomize/blob/master/api/types/var.go#L41
   We cannot use Target directly because it has both `ApiVersion` and `Gvk` (a mistake).
- `Selector`` in: https://github.com/kubernetes-sigs/kustomize/blob/master/api/types/selector.go#L15
 We presumably don't want this because i can select any number of sources.

We don't need a rejector fields in the source selector.

Reply:
resolving this conversation, the PR has been updated to reflect the discussion here